### DISCUSSION TITLE
Gherkin package for automating tasks & tests for non-developer users

### DISCUSSION THREAD

**Discussion by Laurens-makel** [frankframework/frankframework#8044] [General]
Not only developers interact with an IBIS application, especially the web console is used by non-developers and usually those people are the people responsible for acceptance testing.

They often have a defined workflow, involving steps around the web console, i.e:
- Submit message through test-a-pipeline
- Check Ladybug on sending IBIS app
- Check Ladybug on receiving IBIS app
- Copy paste some pre-made query and assert certain values
- Assert there are (no) records in an error store

These actions could be automated by simple API calls towards the FF/Ladybug API's and some additional logic.
By using Gherkin, it becomes possible to translate an API call to natural human language, i.e:

- When I call adapter -name- on -app-name-x- with input message from file -filename-
- Then I expect report -report name- on -app-name-x- to have a checkpoint -checkpoint-name- with value from file -filename-
- And I expect report -report name- on -app-name-y- to have a checkpoint -checkpoint-name- with value from file -filename-

With statements like those, it becomes easier for non-developers to understand a test suite, their results and the impact of failing/succeeding tests.  They might even create a test case themselves as input for a user story, which can be used for applying TDD.

Those tests could involve several different IBIS applications, thus ideal for testing a functional chain which involves multiple adapters from several IBIS applications. 

Besides the testing aspect, I'm also interested in applying the same concept for automating tasks on the console, i.e:

- Given I repeat this task every hour
- When I check the error storage of adapter -name- on -app-name-
- And there are are errors that contain/equal -substring-
- Then I resend those errors

or
- Given the monitor -name- is raised
- Then I check the error storage of adapter -name- on -app-name-
- And there are are errors that contain/equal -substring-
- Then I resend those errors

It could go even further:
- When the error is still present
- Then I turn ladybug on
- And I resend the error again
- Then I download the error report
- And mail this report to -mailbox- 
- And I turn ladybug off

A last thought is that this could be used as an alternative language to describe framework monitor's/alarms.

What are your thoughts?

EDIT, some implementation thoughts:
Once there is an OpenAPI spec of the console API, see #8043, it becomes possible to generate client code using https://github.com/OpenAPITools/openapi-generator/. 

With a mapping file that maps a OpenAPI operation to Gherkin sentences and an extra Moustache template for the generator, it'd be quite easy to generate a Gherkin test utlity package.

This mechanism to generate Gherkin functions could also be applied to the OpenAPI specs generated from adapter configurations or other systems and could be packaged for multiple languages.

**Comment by jjansenvr** [comment]
I think this is a good idea but as we are not using Gherkin at the Frank Framework team it would be a feature you could create? We can help you!

**Reply by Laurens-makel** [reply]
An alternative approach could be to include an annotation which declares available Gherkin sentences for the API endpoint and include it in the FrankDoc.

For API listeners, an additional attribute could be introduced to provide Gherkin sentences which are in the same format the FrankDoc uses and could be included in the FrankDoc.

The same format could be included in the Swagger and based upon the Swagger including the Gherkin format, client code could be generated and matched to the provided Gherkin sentences, this allows the external solution to be framework agnostic, any other solution should be able to provide Gherkin sentences in the same generic format within the Swagger and it should be picked up by the tool.
